src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -150,35 +150,41 @@ void alts_handshaker_client_handle_response(alts_handshaker_client* c,
     cb(TSI_DATA_CORRUPTED, user_data, nullptr, 0, nullptr);
     return;
   }
-  grpc_slice* slice = static_cast<grpc_slice*>(resp->out_frames.arg);
+  upb_strview out_frames = grpc_gcp_HandshakerResp_out_frames(resp);
   unsigned char* bytes_to_send = nullptr;
   size_t bytes_to_send_size = 0;
-  if (slice != nullptr) {
-    bytes_to_send_size = GRPC_SLICE_LENGTH(*slice);
+  if (out_frames.size > 0) {
+    bytes_to_send_size = out_frames.size;
     while (bytes_to_send_size > client->buffer_size) {
       client->buffer_size *= 2;
       client->buffer = static_cast<unsigned char*>(
           gpr_realloc(client->buffer, client->buffer_size));
     }
-    memcpy(client->buffer, GRPC_SLICE_START_PTR(*slice), bytes_to_send_size);
+    memcpy(client->buffer, out_frames.data, bytes_to_send_size);
     bytes_to_send = client->buffer;
   }
   tsi_handshaker_result* result = nullptr;
   if (is_handshake_finished_properly(resp)) {
     alts_tsi_handshaker_result_create(resp, client->is_client, &result);
-    alts_tsi_handshaker_result_set_unused_bytes(result, &client->recv_bytes,
-                                                resp->bytes_consumed);
-  }
-  grpc_status_code code = static_cast<grpc_status_code>(resp->status.code);
+    alts_tsi_handshaker_result_set_unused_bytes(
+        result, &client->recv_bytes,
+        grpc_gcp_HandshakerResp_bytes_consumed(resp));
+  }
+  const grpc_gcp_HandshakerStatus* resp_status =
+      grpc_gcp_HandshakerResp_status(resp);
+  grpc_status_code code = (resp_status != nullptr)
+                              ? static_cast<grpc_status_code>(
+                                    grpc_gcp_HandshakerStatus_code(resp_status))
+                              : GRPC_STATUS_OK;
```

Comment:
The semantics of code (from lines 175 to 178) is slightly different from that of original. That is if `resp` is invalid, it returns `GRPC_STATUS_OK` while in the original code, it directly returns. Could you please double-check?

Reply:
Good catch. I couldn't find the way how to handle when has_status = false with the original code. Which is the better way to handle when status is missing in the message?

Reply:
Is it true in upb if deserialization succeeds (e.g., `resp != nullptr` at line 148), we can assume the validity of all fields in the deserialized message? If that's the case, then I don't think the check at line 175 is necessary. HDYT?

Reply:
It could be null even when deserialization succeeds because status is a nested message so it's valid not to have it, meaning null for upb. In case of nanopb, it has the has_status field to address this but previous code doesn't seem to handle this, right? If not, how can we handle the case of having null status here?

Reply:
Chatted offline and will treat it as a de-serialization failure.